man/man1/git-drover.1
'\" t
.\"     Title: git-drover
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 12/15/2016
.\"    Manual: Chromium depot_tools Manual
.\"    Source: depot_tools f72f1ad
.\"  Language: English
.\"
.TH "GIT\-DROVER" "1" "12/15/2016" "depot_tools f72f1ad" "Chromium depot_tools Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
git-drover \- Apply a commit from the trunk to a release branch, or from one release branch to another\&.
.SH "SYNOPSIS"
.sp
.nf
\fIgit drover\fR \-\-branch <branch>
           (\-\-cherry\-pick <change> | \-\-continue [path_to_workdir] |
            \-\-abort [path_to_workdir])
           [\-\-parent_checkout <path\-to\-existing\-checkout>]
           [\-\-verbose] [\-\-dry\-run]
.fi
.sp
.SH "DESCRIPTION"
.sp
git drover applies a commit to a release branch\&. It creates a new workdir from an existing checkout to avoid downloading a new checkout without affecting the existing checkout\&.
.sp
git drover does not support reverts\&. See the EXAMPLE section for the equivalent sequence of commands to run\&.
.SH "OPTIONS"
.PP
\-\-branch <branch>
.RS 4
The branch to cherry\-pick the commit to\&.
.RE
.PP
\-\-cherry\-pick <commit>
.RS 4
The commit to cherry\-pick\&.
.RE
.PP
\-\-continue [path_to_workdir]
.RS 4
Continue a cherry\-pick that required manual resolution\&. The path to the drover workdir is optional\&. If unspecified, the current directory is used\&.
.RE
.PP
\-\-abort [path_to_workdir]
.RS 4
Abort a cherry\-pick that required manual resolution and clean up its workdir\&. The path to the drover workdir is optional\&. If unspecified, the current directory is used\&.
.RE
.PP
\-\-parent_checkout
.RS 4
The path to the chromium checkout to use as the source for a creating git\-new\-workdir workdir to use for cherry\-picking\&. If unspecified, the current directory is used\&.
.RE
.PP
\-v, \-\-verbose
.RS 4
Enable verbose logging\&.
.RE
.PP
\-\-dry\-run
.RS 4
Skip landing the cherry\-pick\&. Just ensure that the commit can be cherry\-picked into the branch\&.
.RE
.SH "EXAMPLE"
.SS "PREREQUISITES"
.sp
Before working with branches, you must \fIgclient sync \-\-with_branch_heads\fR at least once to fetch the branches\&.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBMerge Example\fR
.RS 4
.sp

.sp
.if n \{\
.RS 4
.\}
.nf
# Here\*(Aqs a commit (from some\&.committer) that we want to \*(Aqdrover\*(Aq\&.
\fB$ git log \-n 1 \-\-pretty=fuller\fR
commit 067e08ebd511040fa9d5350da54112300d9c1801
Author:     some\&.committer <some\&.committer@chromium\&.org>
AuthorDate: Thu Apr 10 08:54:46 2014 +0000
Commit:     some\&.committer <some\&.committer@chromium\&.org>
CommitDate: Thu Apr 10 08:54:46 2014 +0000

    This change needs to go to branch 9999

# Now do the \*(Aqdrover\*(Aq\&.
\fB$ git drover \-\-branch 9999 \-\-cherry\-pick 067e08ebd511040fa9d5350da54112300d9c1801\fR
Going to cherry\-pick
"""
commit 067e08ebd511040fa9d5350da54112300d9c1801
Author: some\&.committer <some\&.committer@chromium\&.org>
Date:   Thu Apr 10 08:54:46 2014 +0000

    This change needs to go to branch 9999
"""
to 9999\&. Continue (y/n)? y

# A cl is uploaded to rietveld, where it can be reviewed before landing\&.

About to land on 9999\&. Continue (y/n)? y
# The cherry\-pick cl is landed on the branch 9999\&.
.fi
.if n \{\
.RE
.\}
.sp
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBMerge with Conflicts Example\fR
.RS 4
.sp

.sp
.if n \{\
.RS 4
.\}
.nf
# Here\*(Aqs a commit (from some\&.committer) that we want to \*(Aqdrover\*(Aq\&.
\fB$ git log \-n 1 \-\-pretty=fuller\fR
commit e4fdcfa1a16c8b7265fb74a20cd029aa75cecaa8
Author:     some\&.committer <some\&.committer@chromium\&.org>
AuthorDate: Thu Apr 10 08:54:46 2014 +0000
Commit:     some\&.committer <some\&.committer@chromium\&.org>
CommitDate: Thu Apr 10 08:54:46 2014 +0000

    This change needs to go to branch 9999

# Now do the \*(Aqdrover\*(Aq\&.
\fB$ git drover \-\-branch 9999 \-\-cherry\-pick e4fdcfa1a16c8b7265fb74a20cd029aa75cecaa8\fR
Going to cherry\-pick
"""
commit e4fdcfa1a16c8b7265fb74a20cd029aa75cecaa8
Author: some\&.committer <some\&.committer@chromium\&.org>
Date:   Thu Apr 10 08:54:46 2014 +0000

    This change needs to go to branch 9999
"""
to 9999\&. Continue (y/n)? y

Error: Patch failed to apply\&.

A workdir for this cherry\-pick has been created in
  /tmp/drover_9999

To continue, resolve the conflicts there and run
  git drover \-\-continue /tmp/drover_9999

To abort this cherry\-pick run
  git drover \-\-abort /tmp/drover_9999

\fB$ pushd /tmp/drover_9999\fR
# Manually resolve conflicts\&.
\fB$ git add path/to/file_with_conflicts\fR
\fB$ popd\fR
\fB$ git drover \-\-continue /tmp/drover_9999\fR

# A cl is uploaded to rietveld, where it can be reviewed before landing\&.

About to land on 9999\&. Continue (y/n)? y
# The cherry\-pick cl is landed on the branch 9999\&.
.fi
.if n \{\
.RE
.\}
.sp
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBRevert Example\fR
.RS 4
.sp

.sp
.if n \{\
.RS 4
.\}
.nf
# Make sure we have the most up\-to\-date branch sources\&.
\fB$ git fetch\fR

# Checkout the branch with the change we want to revert\&.
\fB$ git checkout \-b drover_9999 branch\-heads/9999\fR
Branch drover_9999 set up to track remote ref refs/branch\-heads/9999\&.

# Here\*(Aqs the commit we want to revert\&.
\fB$ git log \-n 1\fR
commit 00b47037e6732455c6b90f8800f4ffe5fbe7326d
Author: some\&.committer <some\&.committer@chromium\&.org>
Date:   Thu Apr 10 08:54:46 2014 +0000

    This change is horribly broken\&.

# Now do the revert\&.
\fB$ git revert 00b47037e6732455c6b90f8800f4ffe5fbe7326d\fR

# That reverted the change and committed the revert\&.
\fB$ git log \-n 1\fR
commit e31cae25b69f773e05ea206848d1479315f2e349
Author: you <you@chromium\&.org>
Date:   Thu Apr 10 09:11:36 2014 +0000

    Revert "This change is horribly broken\&."

    This reverts commit 00b47037e6732455c6b90f8800f4ffe5fbe7326d\&.

# As with old drover, reverts are generally OK to commit without LGTM\&.
\fB$ git cl upload \-r some\&.committer@chromium\&.org \-\-send\-mail\fR
\fB$ git cl land \-\-bypass\-hooks\fR
Using 50% similarity for rename/copy detection\&. Override with \-\-similarity\&.
.fi
.if n \{\
.RE
.\}
.sp
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBManual Merge Example\fR
.RS 4
.sp

.sp
.if n \{\
.RS 4
.\}
.nf
# Make sure we have the most up\-to\-date branch sources\&.
\fB$ git fetch\fR

# Here\*(Aqs a commit (from some\&.committer) that we want to \*(Aqdrover\*(Aq\&.
\fB$ git log \-n 1 \-\-pretty=fuller\fR
commit 87e5c05985d9c2ee235c6b7b5d29cc50f45783db
Author:     some\&.committer <some\&.committer@chromium\&.org>
AuthorDate: Thu Apr 10 08:54:46 2014 +0000
Commit:     some\&.committer <some\&.committer@chromium\&.org>
CommitDate: Thu Apr 10 08:54:46 2014 +0000

    This change needs to go to branch 9999

# Checkout the branch we want to \*(Aqdrover\*(Aq to\&.
\fB$ git checkout \-b drover_9999 branch\-heads/9999\fR
Branch drover_9999 set up to track remote ref refs/branch\-heads/9999\&.

# Now do the \*(Aqdrover\*(Aq\&.
# IMPORTANT!!! Do Not leave off the \*(Aq\-x\*(Aq flag
\fB$ git cherry\-pick \-x 87e5c05985d9c2ee235c6b7b5d29cc50f45783db\fR
[drover_9999 5c879cc] This change needs to go to branch 9999
 Author: some\&.committer <some\&.committer@chromium\&.org>
 Date: Thu Apr 10 08:54:46 2014 +0000
 1 file changed, 1 insertion(+)
 create mode 100644 modified_file

# That took the code authored by some\&.committer and committed it to
# the branch by the person who drovered it (i\&.e\&. you)\&.
\fB$ git log \-n 1 \-\-pretty=fuller\fR
commit 5c879cc29a19045571b08d9b058b8306bf7dee6c
Author:     some\&.committer <some\&.committer@chromium\&.org>
AuthorDate: Thu Apr 10 08:54:46 2014 +0000
Commit:     you <you@chromium\&.org>
CommitDate: Thu Apr 10 09:11:36 2014 +0000

    This change needs to go to branch 9999

    (cherry picked from commit 87e5c05985d9c2ee235c6b7b5d29cc50f45783db)

# Looks good\&. Ship it!
\fB$ git cl upload\fR
# Wait for LGTM or TBR it\&.
\fB$ git cl land\fR
Using 50% similarity for rename/copy detection\&. Override with \-\-similarity\&.
# Or skip the LGTM/TBR and just \*(Aqgit cl land \-\-bypass\-hooks\*(Aq
.fi
.if n \{\
.RE
.\}
.sp
.RE
.SH "SEE ALSO"
.sp
\fBgit-cherry-pick\fR(1), \fBgit-revert\fR(1)
.SH "CHROMIUM DEPOT_TOOLS"
.sp
Part of the chromium \fBdepot_tools\fR(7) suite\&. These tools are meant to assist with the development of chromium and related projects\&. Download the tools from \m[blue]\fBhere\fR\m[]\&\s-2\u[1]\d\s+2\&.
.SH "NOTES"
.IP " 1." 4
here
.RS 4
\%https://chromium.googlesource.com/chromium/tools/depot_tools.git
.RE


NEWS
liblouis NEWS -- history of user-visible changes.  	-*- org -*-

* Noteworthy changes in release 3.1.0 (2016-11-??)
** New features
- Meta data

*** noUndefinedDots mode
Add a noUndefinedDots mode to disable the output of dot numbers when
back-translating undefined patterns. Thanks to James Teh.

When backtranslating input from a braille keyboard cell by cell, it is
desirable to output characters as soon as they are produced.
Similarly, when backtranslating contracted braille, it is desirable to
provide a "guess" to the user of the characters they typed. To achieve
this, liblouis needs to have the ability to produce no text when
indicators (which don't produce a character by themselves) are not
followed by another cell. This works already for indicators liblouis
knows about such as capital sign, number sign, etc., but it does not
work for indicators which are not (and cannot be) specifically defined
as indicators. For example, in UEB, dots 4 5 6 alone produces the text
"\456/". Setting the noUndefinedDots mode suppresses this dot number
output.

*** partialTrans mode
Add a partialTrans mode to specify that back-translation input should
be treated as an incomplete word. Thanks to James Teh.
  
If this mode is set, rules that apply only for complete words or at
the end of a word will not take effect. This is intended to be used
when translating input typed on a braille keyboard to provide a rough
idea to the user of the characters they are typing before the word is
complete.

** Bug fixes
- Fixes implicit declaration of 'pattern_check' thanks to  Reiner Dolp
- Fix a stackoverflow crash on applications with smaller stack size.
  Thanks to Victor Montalvão.
** Other changes
- Improvements to the emacs mode for editing liblouis tables thanks to
  Christian Egli
- Documenting lou_charSize thanks to Reiner Dolp
- Support for relative table paths in the tests. This will make sure
  you always know which table a test actually uses.
- Infrastructure to build windows binaries in a Docker container,
  thanks to Bert Frees
** Braille table improvements
- UEB improvements, thanks to Mike Gray
  - Fixed apostrophe and back translation
  - Added rules for unicode apostrophe handling
- Complete overhaul of Lithuanian 8-dots table, thanks to Rimas
  Kudelis
- New Urdu 6 Dot Grade 1 and 2 Braille tables thanks to Jake Kyle
- Improvements to Italian 8 dots computer braille, thanks to Simone
  Dal Maso.
- New table ~unicode-braille.utb~ that helps to back translate braille
  input to unicode braille output, thanks to Leonard de Ruijter.
- Improvements to the Chinese braille table thanks to Coscell Kao.

** Backwards incompatible changes
None

* Noteworthy changes in release 3.0.0 (2016-07-14)
This is the biggest release of liblouis in years. The major news are
that we now have proper support for UEB and secondly that liblouis is
now licensed under [[https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html][LGPLv2.1+]].

In order to support UEB the internals of liblouis have seen a major
rewrite. New Opcodes have been added to support the requirements for
proper UEB for example for emphasis handling or to handle proper
translation of numbers. Changes to the opcodes are described in the
documentation. Some of these changes are not backwards compatible. All
tables that come with liblouis have been migrated. If you have private
tables look at the section on upgrading from previous versions in the
[[https://github.com/liblouis/liblouis/wiki/Emphasis-Opcodes#upgrade-from-previous-versions][wiki]].

This release also changes the C API. External applications will have
to adapt the way they call liblouis. In particular the typeform
parameter has changed.

The license of the library and most of the tables has been changed to
LGPLv2.1. For a detailed list which tables are still in the process of
migrating the license refer to the [[https://github.com/liblouis/liblouis/wiki/Licensing-of-liblouis-tables][wiki page about the license change]].

** New features
- Numerous features to support UEB properly. Thanks to Mike Gray,
  William Freeman, Davy Kager, Keith Creasy and the American Printing
  House for the Blind for sponsoring this work.
  - support the many emphasis classes needed for UEB.
  - support translation of numbers according to the rules of UEB.
  - Capitalization is now handled just like emphasis.
  - support for numeric mode
- UTF-8 support for all tools thanks to Christian Egli.
- The YAML tests now allow for an optional test description. See the
  documentation for more details.
- Add ~lou_checkTable~ and ~lou_getTypeformForEmphClass~ to the C API
  and to the Python bindings
** Bug fixes
- Stop buffer overrun in ~lou_getProgramPath~, and also free memory
  after usage. Thanks to Michael Curran.
** Other changes
- The license of the library and most of the tables has been changed
  to [[https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html][LGPLv2.1+]].
- Improvements to the test suite:
  - Output is printed to ~stderr~. This helps with locating errors when
    testing with the YAML test suite.
  - typeform is included in output.
** Braille table improvements
- Improved Finish 6-dot braille thanks to Jukka Eerikäinen
- Improvements to the Chinese braille table thanks to Coscell Kao
- Improvements to Mongolian thanks to Tsengel Maidar and Sreeja Param
- Added new Slovak tables based off the official Slovak braille
  standard thanks to Peter Vagner
- Changes to the Norwegian tables. There are now three Norwegian 8-dot
  tables
  - ~no-no-comp8.ctb~: Norwegian 8-dot computer braille table
  - ~no-no-8dot.utb~: Norwegian 8-dot braille table
  - ~no-no-8dot-fallback-6dot-g0.utb~: Norwegian 8-dot braille table
    with uncontracted 6-dot fallback
- Changes to the Dutch tables
  - Conforms better to the standard.
  - ~nl-BE-g1.ctb~ renamed to ~nl-BE-g0.utb~
  - ~nl-NL-g1.ctb~ renamed to ~nl-NL-g0.utb~
- Improvements to Danish tables: Added grade 2 with limited

man/src/git-drover.demo.1.sh
#!/usr/bin/env bash
. git-drover.demo.common.sh

drover_c "This change needs to go to branch 9999"

echo "# Make sure we have the most up-to-date branch sources."
run git fetch
echo
echo "# Here's the commit we want to 'drover'."
run git log -n 1 --pretty=fuller
echo
echo "# Checkout the branch we want to 'drover' to."
run git checkout -b drover_9999 branch-heads/9999
echo
echo "# Now do the 'drover'."
echo "# IMPORTANT!!! Do Not leave off the '-x' flag"
run git cherry-pick -x $(git show-ref -s pick_commit)
echo
echo "# That took the code authored by some.commiter and commited it to the"
echo "# branch by branch.maintainer (us)."
run git log -n 1 --pretty=fuller
echo
echo "# Looks good. Ship it!"
pcommand git cl upload
echo "# Wait for LGTM or TBR it."
run git cl land
echo "# Or skip the LGTM/TBR and just 'git cl land --bypass-hooks'"
